harness/determined/estimator/_estimator_trial.py

```diff
@@ -590,6 +591,9 @@ def run(self) -> None:
             if isinstance(callback, estimator.RunHook):
                 callback.on_trial_close()
 
+        if self.exit_response_func:
```

Comment:
Hm... so if a user has started a thread, which we tell them they can cleanup in callback.on_trial_close()... but if we hit an exception we'll exit in that AssertionError(), and we'll still hang right?

Given what we say the `on_trial_close()` is meant for, we should probably try to get the mechanics such that on_trial_close() gets called no matter what.

And in that case, this `if` statement can become a `check.not_none` and live in the `except det.errors.WorkerFinishedGracefully` section of the `try` statement.

Reply:
If they are running `on_trial_close` on only a portion of the GPUs, and the other GPUs raise an AssertionError we are going to SigKill the process while the callback may be going on. Since we now have the trial timeout, I am going to leave this to be solved as part of [DET-3439](https://determinedai.atlassian.net/browse/DET-3439).